Feature/v2 updates

fivetran-reneeli opened the pull request:
Pull Request
**Are you a current Fivetran customer?** 
<!--- Please tell us your name, title and company -->
Internal
**What change(s) does this PR introduce?** 
<!--- Describe what changes your PR introduces to the package and how to leverage this new feature. -->
This PR applies the Ad Reporting V2 updates:
- Adds in the `ad_squad_hourly_report` and `campaign_hourly_report` source tables and accompanying staging models, and the following field descriptions and tests for the new models in the yml files
- Adds additional fields to existing models that were previously not brought in
- Applies README updates
- Introduces the identifier variable for all source models
- Casts all timestamp fields using dbt_utils.type_timestamp() 

**Did you update the CHANGELOG?** 
<!--- Please update the new package version’s CHANGELOG entry detailing the changes included in this PR. -->
<!--- To select a checkbox you simply need to add an "x" with no spaces between the brackets (eg. [x] Yes). -->
- [x] Yes

**Does this PR introduce a breaking change?**
<!--- Does this PR introduce changes that will cause current package users' jobs to fail or require a `--full-refresh`? -->
<!--- To select a checkbox you simply need to add an "x" with no spaces between the brackets (eg. [x] Yes). -->
- [x] Yes (please provide breaking change details below.)
- [x] No  (please provide an explanation as to how the change is non-breaking below.)

**Did you update the dbt_project.yml files with the version upgrade (please leverage standard semantic versioning)? (In both your main project and integration_tests)** 
<!--- The dbt_project.yml and the integration_tests/dbt_project.yml files contain the version number. Be sure to upgrade it accordingly -->
<!--- To select a checkbox you simply need to add an "x" with no spaces between the brackets (eg. [x] Yes). -->
- [x] Yes

**Is this PR in response to a previously created Bug or Feature Request**
<!--- If an Issue was created it is helpful to track the progress by linking it in the PR. -->
<!--- To select a checkbox you simply need to add an "x" with no spaces between the brackets (eg. [x] Yes). -->
- [ ] Yes, Issue/Feature [link bug/feature number here]
- [x] No 

**How did you test the PR changes?** 
<!--- Proof of testing is required in order for the PR to be approved. -->
<!--- To check a box, remove the space and insert an x in the box (eg. [x] CircleCi). --> 
<!--- To select a checkbox you simply need to add an "x" with no spaces between the brackets (eg. [x] Yes). -->
- [x] CircleCi <!--- CircleCi testing is only applicable to Fivetran employees. --> 
- [ ] Local (please provide additional testing details below)

**Select which warehouse(s) were used to test the PR**
<!--- To check a warehouse remove the space and insert an x in the box (eg. [x] Bigquery). --> 
<!--- To select a checkbox you simply need to add an "x" with no spaces between the brackets (eg. [x] Yes). -->
- [x] BigQuery
- [x] Redshift
- [x] Snowflake
- [x] Postgres
- [x] Databricks
- [ ] Other (provide details below)

**Provide an emoji that best describes your current mood**
<!--- For a complete list of markdown compatible emojis check our this git repo (https://gist.github.com/rxaviers/7360908)  --> 
:dancer:

**Feedback**

We are so excited you decided to contribute to the Fivetran community dbt package! We continue to work to improve the packages and would greatly appreciate your [feedback](https://www.surveymonkey.com/r/DQ7K7WW) on our existing dbt packages or what you'd like to see next.

fivetran-joemarkiewicz reviewed (changes requested):
@fivetran-reneeli thanks so much for putting this PR together. This is very well done and I only have a few very small comments and change requests for you to address. Please mark this for re-review once the comments are addressed!
> FYI we are not going to merge this PR until all other ad reporting packages are updated.

- [ ] Regenerate the docs. I see you generated the run_results.json, but we should regenerate all the docs with these updates.
- [x] Adjust the identifier variable names to be consistent with other packages.
- [ ] Apply the small coding style changes requested within the tmp and staging models.
- [ ] Do you feel there is a desire or a need to add passthrough metrics to the Snapchat report models? Most of our other packages include [these passthrough metrics](https://github.com/fivetran/dbt_google_ads_source/blob/6b5bd6f87270fda66b1ac673ae0a9c282773fe17/models/stg_google_ads__ad_group_stats.sql#L33-L35). If there are other metrics you feel customers would want to include in these models, then I say we add this functionality to the `*_hourly_report` models. Otherwise, we can keep the models as they are.
- [ ] Do you know why for all of our `is_most_recent_record` fields we use `_fivetran_synced` within the partition order by instead of the `updated_at` field within the source? It seems `creative_url_tag_history` seems to be the only one that uses `updated_at`.
- [ ] Add a freshness test to the source that will be consistent with the other ad reporting packages:
```yml
    freshness: 
      warn_after: {count: 48, period: hour}
      error_after: {count: 168, period: hour}
```

- [ ] Apply changes requested within the README.
- [ ] Address any remaining inline comments I have. Be sure to check the "view more comments" in the PR if they are collapsed.

fivetran-joemarkiewicz commented on dbt_project.yml:
```diff
@@ -1,5 +1,5 @@
 name: 'snapchat_ads_source'
-version: '0.3.1'
+version: '0.4.0'
```
Can we change the custom schema on line 18 in this file to be `snapchat_ads_source` to be consistent with our other packages.

fivetran-reneeli commented on models/stg_snapchat.yml:
```diff
@@ -88,15 +175,64 @@ models:
+      - name: impressions
+        description: The number of impressions for an ad in the hour of the record.
+      - name: spend
+        description: The amount of spend in millis for an ad in the hour of the record.
```
Ah, should be microdollars, thanks for noticing

fivetran-reneeli commented on README.md:
```diff
@@ -1,83 +1,110 @@
 vars:
-    snapchat_schema: your_schema_name
-    snapchat_database: your_database_name 
+    snapchat_ads_source_database: your_destination_name
+    snapchat_ads_source_schema: your_schema_name 
```
Hi Joe, by chance did you happen to mean, that "source" isn't included in the variable names? I think "ads" should be included in the names

fivetran-joemarkiewicz replied:
@fivetran-reneeli I agree that `ads` should be in the name. However, the variable itself does not seem to have that naming convention.
https://github.com/fivetran/dbt_snapchat_ads_source/blob/869f7a28af13513120acf0c8114fe9a72d3fa1a3/models/src_snapchat.yml#L5-L6

It looks like the variable names here are just `snapchat_schema` and `snapchat_database`. I think it would be a good idea to include `ads` in the names.

fivetran-reneeli commented:
> * [x]  Do you feel there is a desire or a need to add passthrough metrics to the Snapchat report models? Most of our other packages include [these passthrough metrics](https://github.com/fivetran/dbt_google_ads_source/blob/6b5bd6f87270fda66b1ac673ae0a9c282773fe17/models/stg_google_ads__ad_group_stats.sql#L33-L35). If there are other metrics you feel customers would want to include in these models, then I say we add this functionality to the `*_hourly_report` models. Otherwise, we can keep the models as they are.

Good point! I do see that being relevant so I will add that metric functionality in. I updated the readme as well.

> * [ ]  Do you know why for all of our `is_most_recent_record` fields we use `_fivetran_synced` within the partition order by instead of the `updated_at` field within the source? It seems `creative_url_tag_history` seems to be the only one that uses `updated_at`.

Good question... I glanced at the sample data we had and it looks like it doesn't make much of a difference nor do I think it would when used on real data. I changed the `_fivetran_synced` ones to `updated_at` to make it in line with the other ad packages

fivetran-joemarkiewicz commented on models/stg_snapchat_ads__campaign_hourly_report.sql:
```diff
@@ -0,0 +1,50 @@
+        (spend / 1000000.0) as spend,
+        swipes
+        
+        {% for metric in var('snapchat_ads__campaign_hourly_report_passthrough_metrics', []) %}
+        , {{ metric }}
+        {% endfor %}
```
You will want to add this same logic to the `fields` CTE. Can you do this for each of the models that allow for passthrough metrics. [Here](https://fivetran.slack.com/archives/C02FG7499UK/p1658423688619259) is the example.

If the metric @fivetran-jamie is working on is successful then we will be able to adjust this. But for now we will want to add this into the package.

fivetran-joemarkiewicz commented on macros/get_ad_squad_hourly_report_columns.sql:
```diff
@@ -0,0 +1,49 @@
+{% macro get_ad_squad_hourly_report_columns() %}
```
For each of the macros, you will need to remove the columns that are not used in the final CTE of the staging model. Can you go through the macros and make sure we are only bringing through the relevant fields.

fivetran-joemarkiewicz reviewed (changes requested):
@fivetran-reneeli thanks for going through the changes I requested. There are a few other changes I would like you to apply before approving. It also seems the renaming of the identifier variable from the last review was not applied. Additionally, the `updated_at` field for a few models is included, but not documented. Would you be able to make sure this field is included in our docs.
![image](https://user-images.githubusercontent.com/74217849/180873955-a48294c7-7480-4a28-a878-4b779b523b69.png)

fivetran-reneeli commented on README.md:
```diff
@@ -1,83 +1,112 @@
-# dbt_project.yml
+vars:
+    snapchat_ads_database: your_destination_name
+    snapchat_ads_schema: snapchat_ads_fivetran 
```
Hey @fivetran-joemarkiewicz , wouldn't make more sense to have `snapchat_ads_Schema` as  "snapchat_ads_fivetran" since the paragraph reads:

By default, this package runs using your destination and the `snapchat_ads_source` schema. If this is not where your Snapchat Ads data is (for example, if your Snapchat Ads schema is named `snapchat_ads_fivetran`), add the following configuration to your root `dbt_project.yml` file

fivetran-joemarkiewicz replied:
I definitely see your point. It is quite confusing though that the database variable is then `your_destination_name`. I especially worry about people skimming through and just copying and pasting and being confused here. To keep with standardization across the packages can we change it to `your_schema_name`.

Maybe we can change the paragraph a bit to not say the below snippet is exactly what you would do, but a suggestion of how you override the schema. Thoughts?

fivetran-reneeli replied:
Yup that makes sense to keep to standardization. I'll add clarification to the paragraph!

fivetran-joemarkiewicz commented on CHANGELOG.md:
```diff
@@ -1,3 +1,18 @@
+- Casts all timestamp fields using dbt_utils.type_timestamp()
+- Introduces metric passthrough capability
+
+## 🚨 Breaking Changes 🚨
+- Changes `snapchat_schema` and `snapchat_database` variable names to `snapchat_ads_schema` and `snapchat_ads_database` 
+- Updates model names to prefix with `snapchat_ads`
```
Small request to add breaking changes to the top. This way it is the first thing people see before upgrading.

fivetran-joemarkiewicz reviewed (approved):
@fivetran-reneeli thanks so much for applying these changes! The PR looks good to go (we will hold off merging until we are ready to roll them all out at once).

I only had one super small request to the CHANGELOG to move the breaking changes to the top of the notes for this release.
